crates/subspace-service/src/transaction_pool.rs

```diff
@@ -0,0 +1,554 @@
+        + HeaderBackend<Block>
+        + HeaderMetadata<Block, Error = sp_blockchain::Error>
+        + Send
+        + Sync
+        + 'static,
+    DomainHeader: Header,
+    Client::Api: TaggedTransactionQueue<Block>
+        + SubspaceApi<Block, FarmerPublicKey>
+        + DomainsApi<Block, DomainHeader>,
+{
+    type Block = Block;
+    type Error = sc_transaction_pool::error::Error;
+    type ValidationFuture = ValidationFuture;
+    type BodyFuture = Ready<TxPoolResult<Option<Vec<<Self::Block as BlockT>::Extrinsic>>>>;
+
+    fn validate_transaction(
+        &self,
+        at: &BlockId<Self::Block>,
+        source: TransactionSource,
+        uxt: ExtrinsicFor<Self>,
+    ) -> Self::ValidationFuture {
+        let at = match self.client.block_hash_from_id(at) {
+            Ok(Some(at)) => at,
+            Ok(None) => {
+                let error = sc_transaction_pool::error::Error::BlockIdConversion(format!(
+                    "Failed to convert block id {at} to hash: block not found"
+                ));
+                return Box::pin(async move { Err(error) });
+            }
+            Err(error) => {
+                let error = sc_transaction_pool::error::Error::BlockIdConversion(format!(
+                    "Failed to convert block id {at} to hash: {error}"
+                ));
+                return Box::pin(async move { Err(error) });
+            }
+        };
+
+        let chain_api = self.inner.clone();
+        let client = self.client.clone();
+        let best_block_number = TryInto::<u32>::try_into(client.info().best_number)
+            .expect("Block number will always fit into u32; qed");
+        let diff_in_blocks = self
+            .sync_target_block_number
+            .load(Ordering::Relaxed)
+            .saturating_sub(best_block_number);
+        let slot_probability = self.chain_constants.slot_probability();
+        let fraud_proof_submit_sink = self.fraud_proof_submit_sink.clone();
+        async move {
+            let uxt_validity = chain_api
+                .validate_transaction(&BlockId::Hash(at), source, uxt.clone())
+                .await?;
+
+            if uxt_validity.is_ok() {
+                // Transaction is successfully validated.
+                // If the transaction is `submit_bundle`, then extract the bundle
+                // and check for equivocation.
+                let runtime_api = client.runtime_api();
+                let domains_api_version = runtime_api
+                    .api_version::<dyn DomainsApi<Block, DomainHeader>>(at)
+                    .map_err(|err| {
+                        TxPoolError::RuntimeApi(format!(
+                            "Failed to get `DomainsApi` version for block {at:?}: {err:?}."
+                        ))
+                    })?
+                    // safe to return default version as 1 since there will always be version 1.
+                    .unwrap_or(1);
+                if domains_api_version >= 2 {
```

Comment:
Is this used to keep compatible with gemini-3g? If so please add TODO to remove it before next network.